dipy/reconst/msdki.py

```diff
@@ -56,6 +57,163 @@ def mean_signal_bvalue(data, gtab, bmag=None):
+    Parameters
+    ----------
+    f : ndarray ([X, Y, Z, ...])
+        ndarray containing the axonal volume fraction estimate.
+
+    Returns
+    -------
+    dkdf : ndarray(nub)
+        Mean signal kurtosis differential
+    msk : float
+        Measured mean signal kurtosis.
+
+    Notes
+    -----
+    This function corresponds to the differential of equations 17 of [1]_.
+    This function is applicable to both _msk_from_awf and _msk_from_awf_error.
+
+    References
+    ----------
+    .. [1] Neto Henriques R, Jespersen SN, Shemesh N (2019). Microscopic
+           anisotropy misestimation in spherical‐mean single diffusion
+           encoding MRI. Magnetic Resonance in Medicine (In press).
+           doi: 10.1002/mrm.27606
+    """
+    F = 216*f - 504 * f**2 + 504 * f**3 - 180 * f**4  # Numerator
+    G = 135 - 360*f + 420 * f**2 - 240 * f**3 + 60 * f**4  # Denominator
+
+    dF = 216 - 1008 * f + 1512 * f**2 - 720 * f**3  # Num. differential
+    dG = -360 + 840 * f - 720 * f**2 + 240 * f**3  # Den. differential
+
+    return (G * dF - F * dG) / (G ** 2)
+
+
+def awf_from_msk(msk, mask=None):
+    """
+    Computes the axonal water fraction from the mean signal kurtosis
+    assuming the 2-compartmental spherical mean technique model [1]_, [2]_
+
+    Parameters
+    ----------
+    msk : ndarray ([X, Y, Z, ...])
+        Mean signal kurtosis (msk)
+    mask : ndarray, optional
+        A boolean array used to mark the coordinates in the data that should be
+        analyzed that has the same shape of the msdki parameters
+
+    Returns
+    -------
+    smt2f : ndarray ([X, Y, Z, ...])
+        ndarray containing the axonal volume fraction estimate.
+
+    Notes
+    -----
+    Computes the axonal water fraction from the mean signal kurtosis
+    MSK using equation 17 of [1]_
+
+    References
+    ----------
+    .. [1] Neto Henriques R, Jespersen SN, Shemesh N (2019). Microscopic
+           anisotropy misestimation in spherical‐mean single diffusion
+           encoding MRI. Magnetic Resonance in Medicine (In press).
+           doi: 10.1002/mrm.27606
+    .. [2] Kaden E, Kelm ND, Carson RP, et al. (2016) Multi‐compartment
+           microscopic diffusion imaging. Neuroimage 139:346–359.
+    """
+    awf = np.zeros(msk.shape)
+
+    # Prepare mask
+    if mask is None:
+        mask = np.ones(msk.shape, dtype=bool)
+    else:
+        if mask.shape != msk.shape:
+            raise ValueError("Mask is not the same shape as data.")
+        mask = np.array(mask, dtype=bool, copy=False)
+
+    # looping voxels
+    index = ndindex(mask.shape)
+    for v in index:
+        # Skip if out of mask
+        if not mask[v]:
+            continue
+
+        if msk[v] > 2.4:
+            awf[v] = 1
+        elif msk[v] < 0:
+            awf[v] = 0
+        else:
+            mski = msk[v]
+            fini = mski / 2.4  # Initial guess based on linear assumption
+            awf[v] = opt.fsolve(_msk_from_awf_error, fini, args=(mski,),
```

Comment:
Do you want to catch and raise a warning here if this does not converge properly? Users will see something like:

```
Users/arokem/.virtualenvs/dipy/lib/python3.7/site-packages/scipy/optimize/minpack.py:162: RuntimeWarning: The iteration is not making good progress, as measured by the 
  improvement from the last ten iterations.
```

Which is pretty cryptic.